frontend/src/lib/components/PropertyGroupFilters/propertyGroupFilterLogic.ts

```diff
@@ -0,0 +1,92 @@
+import { PropertyGroupFilterLogicProps } from 'lib/components/PropertyFilters/types'
+import { AndOr } from '../PropertyGroupFilters/PropertyGroupFilters'
+
+import { propertyGroupFilterLogicType } from './propertyGroupFilterLogicType'
+
+export const propertyGroupFilterLogic = kea<propertyGroupFilterLogicType>({
+    path: (key) => ['lib', 'components', 'PropertyGroupFilters', 'propertyGroupFilterLogic', key],
+    props: {} as PropertyGroupFilterLogicProps,
+    key: (props) => props.pageKey,
+
+    actions: () => ({
+        update: (propertyGroupIndex?: number) => ({ propertyGroupIndex }),
+        setFilters: (filters: AndOrPropertyFilter) => ({ filters }),
+        removeFilterGroup: (filterGroup: number) => ({ filterGroup }),
+        setPropertyGroupsType: (type: AndOr) => ({ type }),
+        setPropertyFilters: (properties, index: number) => ({ properties, index }),
+        setPropertyGroupType: (type: AndOr, index: number) => ({ type, index }),
+        addFilterGroup: true,
+    }),
+
+    reducers: ({ props }) => ({
+        filters: [
+            props.propertyFilters ? parsePropertyGroups(props.propertyFilters) : ({} as AndOrPropertyFilter),
+            {
+                setFilters: (_, { filters }) => filters,
+                addFilterGroup: (state) => {
+                    if (!state.groups) {
+                        return {
+                            groups: [
+                                {
+                                    type: AndOr.AND,
+                                    groups: [{}],
+                                },
+                            ],
+                        }
+                    }
+                    const groupsCopy = { ...state }
+                    groupsCopy.groups.push({ type: AndOr.AND, groups: [{}] })
+
+                    if (groupsCopy.groups.length > 1 && !groupsCopy.type) {
+                        groupsCopy.type = AndOr.AND
+                    }
+                    return groupsCopy
+                },
+                removeFilterGroup: (state, { filterGroup }) => {
+                    const newState = { ...state, groups: state.groups.filter((_, idx: number) => idx !== filterGroup) }
+                    if (newState.groups.length <= 1) {
+                        return { groups: newState.groups }
+                    }
+                    return newState
+                },
+                setPropertyGroupsType: (state, { type }) => {
+                    return { ...state, type }
+                },
+                setPropertyFilters: (state, { properties, index }) => {
+                    const newState = { ...state }
+                    newState.groups[index].groups = properties
+                    // removes entire property group if no properties in groups
+                    if (newState.groups[index].groups.length === 0) {
+                        newState.groups = newState.groups.filter((_, i: number) => i !== index)
+                    }
```

Comment:
Same here and below. `groups` is a reference to the same location in memory, for both `newState` and `state`.

Reply:
So I should be remapping everything inside of each nested group object in the array 😭 yeah I figured this needed deep cloning, thanks for catching it!!

Reply:
I'm using a small clone package to do this, hope that's fine!!